src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java

```diff
@@ -0,0 +1,164 @@
+        anotherUser.setPassword("anotherUserPassword");
+        user.setId(1L);
+        user.setAge(23);
+        user.setLogin("uniqueLogin");
+        user.setPassword("uniquePassword");
+    }
+
+    @AfterEach
+    void tearDown() {
+        storageDao = new StorageDaoImpl();
+    }
+
+    @Test
+    void register_NullUser_NotOk() {
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(null);
+        });
+    }
+
+    @Test
+    void register_NullId_NotOk() {
+        user.setId(null);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_NullAge_NotOk() {
+        user.setAge(null);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_NullLogin_NotOk() {
+        user.setLogin(null);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_NullPassword_NotOk() {
+        user.setPassword(null);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_NegativeId_NotOk() {
+        user.setId(-1L);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_invalidPasswordLength_NotOk() {
+        user.setPassword("12345");
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+        anotherUser.setPassword("1234");
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(anotherUser);
+        });
+    }
+
+    @Test
+    void register_SameUser_NotOk() {
+        User sameUser = user;
+        storageDao.add(user);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(sameUser);
+        });
+    }
+
+    @Test
+    void register_UsersWithSameLogin_NotOk() {
+        storageDao.add(user);
+        anotherUser.setLogin(user.getLogin());
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(anotherUser);
+        });
+        assertEquals(storageDao.get(user.getLogin()), user);
+    }
+
+    @Test
+    void register_AgeLessThanNeeded_NotOk() {
+        user.setAge(17);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_UserNegativeAge_NotOk() {
+        user.setAge(-25);
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(user);
+        });
+    }
+
+    @Test
+    void register_TwoDifferentUsers_Ok() {

```

Comment:
I'm not sure what this checks and whether we need it

Reply:
I was trying to test normal cases, when we add to different users in storage and checking that they are really different

Reply:
Interesting, when i removed that test case, test coverage icreased to 95%